flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/AggregateOperationFactory.java

```diff
@@ -501,16 +493,40 @@ protected Void defaultMethod(ResolvedExpression expression) {
 		}
 	}
 
-	private class GroupingExpressionValidator extends ResolvedExpressionDefaultVisitor<Void> {
+	private static class IsKeyTypeChecker implements DataTypeVisitor<Boolean> {
+
 		@Override
-		protected Void defaultMethod(ResolvedExpression expression) {
-			TypeInformation<?> groupingType = expressionBridge.bridge(expression).resultType();
+		public Boolean visit(AtomicDataType atomicDataType) {
```

Comment:
nit: I think we could still simplify this code by just using a `LogicalTypeDefaultVisitor` and iterate over `LogicalType#getChildren`?

Reply:
Not really. `LegacyTypeInformationType` goes through default method. The same happens with `LegacyTypeInformationType`. We would have exact same code in the `visit` method of the `LogicalTypeDefaultVisitor`